File: credit_card_validator.py
card_number = "468368746945295"


def get_card_type(card_number):
    card_type = ""

    if card_number[0] == "4":
        card_type = "Visa Cards"
    elif card_number[0] == "5":
        card_type = "MasterCard"
    elif card_number[0] == "3" and card_number[1] == "7":
        card_type = "American Express Cards"
    elif card_number[0]:
        card_type = "Discover cards"
    else:
        card_type = "Invalid Card"

    return card_type


def get_card_length(card_number):
    length = len(card_number)
    
    return length
def lhun_algorithm(card_number):
    length = len(card_number)
    digits = []
    for character in card_number:
        digits.append(int(character))

    sum_of_even_places = 0
    sum_of_odd_places = 0


    counter = 0 


    for index in range(length - 1, -1, -1):
        current_digit = digits[index]

        if counter % 2 == 1:
            doubled = current_digit * 2
            

            if doubled > 9:
                first_digit = doubled // 10
                second_digit = doubled % 10
                doubled = first_digit + second_digit
                
            sum_of_even_places += doubled
        else:
            sum_of_odd_places += current_digit
            
        counter = counter + 1


    total_sum = sum_of_even_places + sum_of_odd_places

    return total_sum

def check_card_validity(card_number):
    card_type = get_card_type(card_number)
    length = get_card_length(card_number)
    luhn_algotithm = lhun_algorithm(card_number)

    validity = "Invalid"

    if length >= 13 and length <= 16 and card_type != "Invalid Card" and luhn_algotithm % 10 == 0:
        validity = "Valid"
    return validity


print("**************************************************")
print("Credit Card Type:", get_card_type(card_number))
print("Credit Card Number:", card_number)
print("Credit Card Digit Length:", get_card_length(card_number))
print("Credit Card Validity Status:", check_card_validity)
print("**************************************************")

File: CreditCardValidator.java
public class CreditCardValidator{

    public static void main(String[] args){


        String cardNumber = "468368746945295";

        System.out.println("**************************************************");
        System.out.println("Credit Card Type: " + getCardType(cardNumber));
        System.out.println("Credit Card Number: " + cardNumber);
        System.out.println("Credit Card Digit Length: " + getCardLength(cardNumber));
        System.out.println("Credit Card Validity Status: " + getCardValidity(cardNumber));
        System.out.println("**************************************************");

    }

    public static String getCardType(String cardNumber){
        String cardType = "";

        if (cardNumber.charAt(0) == '4'){
            cardType = "Visa Card";
        }
        else if (cardNumber.charAt(0) == '5'){
            cardType = "MasterCard";
        }
        else if (cardNumber.charAt(0) == '3' && cardNumber.charAt(1) == '7'){
            cardType = "American Express Card";
        }
        else if (cardNumber.charAt(0) == '6'){
            cardType = "Discover Card";
        }
        else{
            cardType = "Invalid Card";
        }
        
        return cardType;

    }

    public static int getCardLength(String cardNumber){
        int length = cardNumber.length();
        return length;
    }

    public static int lhunAlgorithm(String cardNumber){

        int cardLength = getCardLength(cardNumber);

        int[] digits = new int[cardLength];
        for (int index = 0; index < cardLength; index++){
            digits[index] = cardNumber.charAt(index) - '0';
        }


        int sumOfEvenPlaces = 0;
        int sumOfOddPlaces = 0;


        int counter = 0;
        int currentDigit = 0;
        for (int index = cardLength - 1; index >= 0; index--){
            currentDigit = digits[index];

            if (counter % 2 == 1){
                int doubled = currentDigit * 2;

                if (doubled > 9){
                    int firstDigit = doubled / 10;
                    int secondDigit = doubled % 10;
                    doubled = firstDigit + secondDigit;
                }
                sumOfEvenPlaces += doubled;
            }
            else{
                sumOfOddPlaces += currentDigit;
            }
            
            counter += 1;
        }


        int totalSum = sumOfEvenPlaces + sumOfOddPlaces;

        return totalSum;
    }

    public static String getCardValidity(String cardNumber){
        String cardType = getCardType(cardNumber);
        int cardLength = getCardLength(cardNumber);
        int luhnAlgorithm = lhunAlgorithm(cardNumber);

        String validity = "Invalid";

        if(cardLength >= 13 && cardLength <= 16 && cardType != "Invalid Card" && luhnAlgorithm % 10 == 0){
            validity = "Valid";
        }

        return validity;
    }
}


